// purify-ts@1.2.0/Codec.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseError = exports.map = exports.intersect = exports.date = exports.tuple = exports.nonEmptyList = exports.maybe = exports.lazy = exports.exactly = exports.record = exports.array = exports.oneOf = exports.enumeration = exports.unknown = exports.boolean = exports.nullable = exports.optional = exports.nullType = exports.number = exports.string = exports.Codec = void 0;
var Either_1 = require("./Either");
var Function_1 = require("./Function");
var Maybe_1 = require("./Maybe");
var NonEmptyList_1 = require("./NonEmptyList");
var isEmptySchema = function (schema) {
    return Object.keys(schema).length === 0;
};
var isObject = function (obj) {
    return typeof obj === 'object' && obj !== null && !Array.isArray(obj);
};
var reportError = function (expectedType, input) {
    var receivedString = '';
    switch (typeof input) {
        case 'undefined':
            receivedString = 'undefined';
            break;
        case 'object':
            receivedString =
                input === null
                    ? 'null'
                    : Array.isArray(input)
                        ? 'an array with value ' + JSON.stringify(input)
                        : 'an object with value ' + JSON.stringify(input);
            break;
        case 'boolean':
            receivedString = 'a boolean';
            break;
        case 'symbol':
            receivedString = 'a symbol';
            break;
        case 'function':
            receivedString = 'a function';
            break;
        case 'bigint':
            receivedString = "a bigint with value ".concat(input.toString());
    }
    receivedString =
        receivedString || "a ".concat(typeof input, " with value ").concat(JSON.stringify(input));
    return "Expected ".concat(expectedType, ", but received ").concat(receivedString);
};
var removeOneOfWithSingleElement = function (schema) {
    var _a;
    var schemaKeys = Object.keys(schema);
    if (schemaKeys.length === 1 &&
        ((_a = schema.oneOf) === null || _a === void 0 ? void 0 : _a.length) === 1 &&
        typeof schema.oneOf[0] === 'object') {
        Object.assign(schema, schema.oneOf[0]);
        delete schema.oneOf;
    }
    return schema;
};
var flattenNestedOneOf = function (schema) {
    var _a;
    if (Array.isArray(schema.oneOf)) {
        for (var i = 0; i < schema.oneOf.length; i++) {
            var e = schema.oneOf[i];
            if (typeof e === 'object' && e.oneOf) {
                schema.oneOf.splice(i, 1);
                (_a = schema.oneOf).push.apply(_a, __spreadArray([], __read(e.oneOf), false));
                return optimizeSchema(schema);
            }
        }
    }
    return schema;
};
var optimizeSchema = function (schema) {
    flattenNestedOneOf(schema);
    removeOneOfWithSingleElement(schema);
    return schema;
};
exports.Codec = {
    /** Creates a codec for any JSON object */
    interface: function (properties) {
        var keys = Object.keys(properties);
        var decode = function (input) {
            var e_1, _a;
            if (!isObject(input)) {
                return (0, Either_1.Left)(reportError('an object', input));
            }
            var result = {};
            try {
                for (var keys_1 = __values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                    var key = keys_1_1.value;
                    if (!input.hasOwnProperty(key) &&
                        !properties[key]._isOptional) {
                        return (0, Either_1.Left)("Problem with property \"".concat(key, "\": it does not exist in received object ").concat(JSON.stringify(input)));
                    }
                    var decodedProperty = properties[key].decode(input[key]);
                    if (decodedProperty.isLeft()) {
                        return (0, Either_1.Left)("Problem with the value of property \"".concat(key, "\": ").concat(decodedProperty.extract()));
                    }
                    var value = decodedProperty.extract();
                    if (value !== undefined) {
                        result[key] = value;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return (0, Either_1.Right)(result);
        };
        var encode = function (input) {
            var e_2, _a;
            var result = {};
            try {
                for (var keys_2 = __values(keys), keys_2_1 = keys_2.next(); !keys_2_1.done; keys_2_1 = keys_2.next()) {
                    var key = keys_2_1.value;
                    result[key] = properties[key].encode(input[key]);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (keys_2_1 && !keys_2_1.done && (_a = keys_2.return)) _a.call(keys_2);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return result;
        };
        return {
            decode: decode,
            encode: encode,
            unsafeDecode: function (input) { return decode(input).mapLeft(Error).unsafeCoerce(); },
            schema: function () {
                return keys.reduce(function (acc, key) {
                    var isOptional = properties[key]._isOptional;
                    if (!isOptional) {
                        acc.required.push(key);
                    }
                    acc.properties[key] = optimizeSchema(properties[key].schema());
                    return acc;
                }, {
                    type: 'object',
                    properties: {},
                    required: []
                });
            }
        };
    },
    /** Creates a codec for any type, you can add your own deserialization/validation logic in the decode argument */
    custom: function (_a) {
        var decode = _a.decode, encode = _a.encode, schema = _a.schema;
        return {
            decode: decode,
            encode: encode,
            unsafeDecode: function (input) { return decode(input).mapLeft(Error).unsafeCoerce(); },
            schema: schema !== null && schema !== void 0 ? schema : (function () { return ({}); })
        };
    }
};
/** A codec for any string value. Most of the time you will use it to implement an interface codec (see the Codec#interface example above). Encoding a string acts like the identity function */
exports.string = exports.Codec.custom({
    decode: function (input) {
        return typeof input === 'string'
            ? (0, Either_1.Right)(input)
            : (0, Either_1.Left)(reportError('a string', input));
    },
    encode: Function_1.identity,
    schema: function () { return ({ type: 'string' }); }
});
/** A codec for any number value. This includes anything that has a typeof number - NaN, Infinity etc. Encoding a number acts like the identity function */
exports.number = exports.Codec.custom({
    decode: function (input) {
        return typeof input === 'number'
            ? (0, Either_1.Right)(input)
            : (0, Either_1.Left)(reportError('a number', input));
    },
    encode: Function_1.identity,
    schema: function () { return ({ type: 'number' }); }
});
/** A codec for null only */
exports.nullType = exports.Codec.custom({
    decode: function (input) {
        return input === null ? (0, Either_1.Right)(input) : (0, Either_1.Left)(reportError('a null', input));
    },
    encode: Function_1.identity,
    schema: function () { return ({ type: 'null' }); }
});
var undefinedType = exports.Codec.custom({
    decode: function (input) {
        return input === undefined
            ? (0, Either_1.Right)(input)
            : (0, Either_1.Left)(reportError('an undefined', input));
    },
    encode: Function_1.identity
});
var optional = function (codec) {
    return (__assign(__assign({}, (0, exports.oneOf)([codec, undefinedType])), { schema: codec.schema, _isOptional: true }));
};
exports.optional = optional;
/** A codec for a value T or null. Keep in mind if you use `nullable` inside `Codec.interface` the property will still be required */
var nullable = function (codec) {
    return (0, exports.oneOf)([codec, exports.nullType]);
};
exports.nullable = nullable;
/** A codec for a boolean value */
exports.boolean = exports.Codec.custom({
    decode: function (input) {
        return typeof input === 'boolean'
            ? (0, Either_1.Right)(input)
            : (0, Either_1.Left)(reportError('a boolean', input));
    },
    encode: Function_1.identity,
    schema: function () { return ({ type: 'boolean' }); }
});
/** A codec that can never fail, but of course you get no type information. Encoding an unknown acts like the identity function */
exports.unknown = exports.Codec.custom({
    decode: Either_1.Right,
    encode: Function_1.identity,
    schema: function () { return ({}); }
});
/** A codec for a TypeScript enum */
var enumeration = function (e) {
    var enumValues = Object.values(e);
    return exports.Codec.custom({
        decode: function (input) {
            return (0, exports.oneOf)([exports.string, exports.number])
                .decode(input)
                .chain(function (x) {
                var enumIndex = enumValues.indexOf(x);
                return enumIndex !== -1
                    ? (0, Either_1.Right)(enumValues[enumIndex])
                    : (0, Either_1.Left)(reportError('an enum member', input));
            });
        },
        encode: Function_1.identity,
        schema: function () { return ({ enum: enumValues }); }
    });
};
exports.enumeration = enumeration;
/** A codec combinator that receives a list of codecs and runs them one after another during decode and resolves to whichever returns Right or to Left if all fail */
var oneOf = function (codecs) {
    return exports.Codec.custom({
        decode: function (input) {
            var e_3, _a;
            var errors = [];
            try {
                for (var codecs_1 = __values(codecs), codecs_1_1 = codecs_1.next(); !codecs_1_1.done; codecs_1_1 = codecs_1.next()) {
                    var codec = codecs_1_1.value;
                    var res = codec.decode(input);
                    if (res.isRight()) {
                        return res;
                    }
                    else {
                        errors.push(res.extract());
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (codecs_1_1 && !codecs_1_1.done && (_a = codecs_1.return)) _a.call(codecs_1);
                }
                finally { if (e_3) throw e_3.error; }
            }
            return (0, Either_1.Left)("One of the following problems occured: ".concat(errors
                .map(function (err, i) { return "(".concat(i, ") ").concat(err); })
                .join(', ')));
        },
        encode: function (input) {
            var e_4, _a;
            var _loop_1 = function (codec) {
                var res = Either_1.Either.encase(function () { return codec.encode(input); })
                    .mapLeft(function (_) { return ''; })
                    .chain(codec.decode);
                if (res.isRight()) {
                    return { value: codec.encode(input) };
                }
            };
            try {
                for (var codecs_2 = __values(codecs), codecs_2_1 = codecs_2.next(); !codecs_2_1.done; codecs_2_1 = codecs_2.next()) {
                    var codec = codecs_2_1.value;
                    var state_1 = _loop_1(codec);
                    if (typeof state_1 === "object")
                        return state_1.value;
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (codecs_2_1 && !codecs_2_1.done && (_a = codecs_2.return)) _a.call(codecs_2);
                }
                finally { if (e_4) throw e_4.error; }
            }
            return input;
        },
        schema: function () { return ({ oneOf: codecs.map(function (x) { return x.schema(); }) }); }
    });
};
exports.oneOf = oneOf;
/** A codec for an array */
var array = function (codec) {
    return exports.Codec.custom({
        decode: function (input) {
            if (!Array.isArray(input)) {
                return (0, Either_1.Left)(reportError('an array', input));
            }
            else {
                var result = [];
                for (var i = 0; i < input.length; i++) {
                    var decoded = codec.decode(input[i]);
                    if (decoded.isRight()) {
                        result.push(decoded.extract());
                    }
                    else {
                        return (0, Either_1.Left)("Problem with the value at index ".concat(i, ": ").concat(decoded.extract()));
                    }
                }
                return (0, Either_1.Right)(result);
            }
        },
        encode: function (input) { return input.map(codec.encode); },
        schema: function () { return ({
            type: 'array',
            items: codec.schema()
        }); }
    });
};
exports.array = array;
var numberString = exports.Codec.custom({
    decode: function (input) {
        return exports.string
            .decode(input)
            .chain(function (x) {
            return isFinite(+x) ? (0, Either_1.Right)(x) : (0, Either_1.Left)(reportError('a number', input));
        });
    },
    encode: Function_1.identity,
    schema: exports.number.schema
});
/** A codec for an object without specific properties, its restrictions are equivalent to the Record<K, V> type so you can only check for number and string keys */
var record = function (keyCodec, valueCodec) {
    return exports.Codec.custom({
        decode: function (input) {
            var e_5, _a;
            var result = {};
            var keyCodecOverride = keyCodec === exports.number ? numberString : keyCodec;
            if (!isObject(input)) {
                return (0, Either_1.Left)(reportError('an object', input));
            }
            try {
                for (var _b = __values(Object.keys(input)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    if (input.hasOwnProperty(key)) {
                        var decodedKey = keyCodecOverride.decode(key);
                        var decodedValue = valueCodec.decode(input[key]);
                        if (decodedKey.isRight() && decodedValue.isRight()) {
                            result[decodedKey.extract()] = decodedValue.extract();
                        }
                        else if (decodedKey.isLeft()) {
                            return (0, Either_1.Left)("Problem with key type of property \"".concat(key, "\": ").concat(decodedKey.extract()));
                        }
                        else if (decodedValue.isLeft()) {
                            return (0, Either_1.Left)("Problem with the value of property \"".concat(key, "\": ").concat(decodedValue.extract()));
                        }
                    }
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_5) throw e_5.error; }
            }
            return (0, Either_1.Right)(result);
        },
        encode: function (input) {
            var result = {};
            for (var key in input) {
                if (input.hasOwnProperty(key)) {
                    result[keyCodec.encode(key)] = valueCodec.encode(input[key]);
                }
            }
            return result;
        },
        schema: function () { return ({
            type: 'object',
            additionalProperties: valueCodec.schema()
        }); }
    });
};
exports.record = record;
/** A codec that only succeeds decoding when the value is exactly what you've constructed the codec with */
var exactly = function () {
    var expectedValues = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        expectedValues[_i] = arguments[_i];
    }
    return exports.Codec.custom({
        decode: function (input) {
            return expectedValues.includes(input)
                ? (0, Either_1.Right)(input)
                : (0, Either_1.Left)(reportError(expectedValues.map(function (x) { return JSON.stringify(x); }).join(', '), input));
        },
        encode: Function_1.identity,
        schema: function () { return ({
            oneOf: expectedValues.map(function (value) { return ({
                type: typeof value,
                enum: [value]
            }); })
        }); }
    });
};
exports.exactly = exactly;
/** A special codec used when dealing with recursive data structures, it allows a codec to be recursively defined by itself */
var lazy = function (getCodec) {
    return exports.Codec.custom({
        decode: function (input) { return getCodec().decode(input); },
        encode: function (input) { return getCodec().encode(input); },
        schema: function () { return ({
            $comment: 'Lazy codecs are not supported when generating a JSON schema'
        }); }
    });
};
exports.lazy = lazy;
/** A codec for purify's Maybe type. Encode runs Maybe#toJSON, which effectively returns the value inside if it's a Just or undefined if it's Nothing */
var maybe = function (codec) {
    var baseCodec = exports.Codec.custom({
        decode: function (input) {
            return Maybe_1.Maybe.fromNullable(input).caseOf({
                Just: function (x) { return codec.decode(x).map(Maybe_1.Just); },
                Nothing: function () { return (0, Either_1.Right)(Maybe_1.Nothing); }
            });
        },
        encode: function (input) { return input.map(codec.encode).orDefault(undefined); },
        schema: function () {
            return isEmptySchema(codec.schema())
                ? {}
                : { oneOf: [codec.schema(), { type: 'null' }] };
        }
    });
    return __assign(__assign({}, baseCodec), { _isOptional: true });
};
exports.maybe = maybe;
/** A codec for purify's NEL type */
var nonEmptyList = function (codec) {
    var arrayCodec = (0, exports.array)(codec);
    return exports.Codec.custom({
        decode: function (input) {
            return arrayCodec
                .decode(input)
                .chain(function (x) {
                return NonEmptyList_1.NonEmptyList.fromArray(x).toEither("Expected an array with one or more elements, but received an empty array");
            });
        },
        encode: arrayCodec.encode,
        schema: function () { return (__assign(__assign({}, arrayCodec.schema()), { minItems: 1 })); }
    });
};
exports.nonEmptyList = nonEmptyList;
/** The same as the array decoder, but accepts a fixed amount of array elements and you can specify each element type, much like the tuple type */
var tuple = function (codecs) {
    return exports.Codec.custom({
        decode: function (input) {
            if (!Array.isArray(input)) {
                return (0, Either_1.Left)(reportError('an array', input));
            }
            else if (codecs.length !== input.length) {
                return (0, Either_1.Left)("Expected an array of length ".concat(codecs.length, ", but received an array with length of ").concat(input.length));
            }
            else {
                var result = [];
                for (var i = 0; i < codecs.length; i++) {
                    var decoded = codecs[i].decode(input[i]);
                    if (decoded.isRight()) {
                        result.push(decoded.extract());
                    }
                    else {
                        return (0, Either_1.Left)("Problem with the value at index ".concat(i, ": ").concat(decoded.extract()));
                    }
                }
                return (0, Either_1.Right)(result);
            }
        },
        encode: function (input) { return input.map(function (x, i) { return codecs[i].encode(x); }); },
        schema: function () { return ({
            type: 'array',
            items: codecs.map(function (x) { return x.schema(); }),
            additionalItems: false,
            minItems: codecs.length,
            maxItems: codecs.length
        }); }
    });
};
exports.tuple = tuple;
/** A codec for a parsable date string, on successful decoding it resolves to a Date object. The validity of the date string during decoding is decided by the browser implementation of Date.parse. Encode runs toISOString on the passed in date object */
exports.date = exports.Codec.custom({
    decode: function (input) {
        return exports.string
            .decode(input)
            .mapLeft(function (err) { return "Problem with date string: ".concat(err); })
            .chain(function (x) {
            return Number.isNaN(Date.parse(x))
                ? (0, Either_1.Left)('Expected a valid date string, but received a string that cannot be parsed')
                : (0, Either_1.Right)(new Date(x));
        });
    },
    encode: function (input) { return input.toISOString(); },
    schema: function () { return ({ type: 'string', format: 'date-time' }); }
});
/** Creates an intersection between two codecs. If the provided codecs are not for an object, the second decode result will be returned */
var intersect = function (t, u) {
    return exports.Codec.custom({
        decode: function (input) {
            var et = t.decode(input);
            if (et.isLeft()) {
                return et;
            }
            var eu = u.decode(input);
            if (eu.isLeft()) {
                return eu;
            }
            var valuet = et.extract();
            var valueu = eu.extract();
            return isObject(valuet) && isObject(valueu)
                ? (0, Either_1.Right)(Object.assign(valuet, valueu))
                : (0, Either_1.Right)(valueu);
        },
        encode: function (input) {
            var valuet = t.encode(input);
            var valueu = u.encode(input);
            return isObject(valuet) && isObject(valueu)
                ? Object.assign(valuet, valueu)
                : valueu;
        },
        schema: function () { return ({ allOf: [t, u].map(function (x) { return x.schema(); }) }); }
    });
};
exports.intersect = intersect;
/** A codec for the built-in Map type */
var map = function (keyCodec, valueCodec) {
    return exports.Codec.custom({
        decode: function (input) {
            return (0, exports.array)((0, exports.tuple)([keyCodec, valueCodec]))
                .decode(input)
                .map(function (pairs) { return new Map(pairs); });
        },
        encode: function (input) {
            return Array.from(input.entries()).map(function (_a) {
                var _b = __read(_a, 2), k = _b[0], v = _b[1];
                return [
                    keyCodec.encode(k),
                    valueCodec.encode(v)
                ];
            });
        },
        schema: function () { return ({
            type: 'array',
            items: {
                type: 'array',
                items: [keyCodec.schema(), valueCodec.schema()],
                additionalItems: false,
                minItems: 2,
                maxItems: 2
            }
        }); }
    });
};
exports.map = map;
var oneofRegex = /^(One of the following problems occured:)\s/;
var oneOfCounterRegex = /\(\d\)\s/;
var oneOfSeparatorRegex = /\, (?=\()/g;
var failureRegex = /^(Expected ).+(, but received )/;
var failureReceivedSeparator = ' with value';
var missingPropertyMarker = 'Problem with property "';
var badPropertyMarker = 'Problem with the value of property "';
var badPropertyKeyMarker = 'Problem with key type of property "';
var dateFailureMarket = 'Problem with date string: ';
var indexMarker = 'Problem with the value at index ';
var expectedTypesMap = {
    'an object': 'object',
    'a number': 'number',
    'a string': 'string',
    'an undefined': 'undefined',
    'a boolean': 'boolean',
    'an array': 'array',
    'a null': 'null',
    'an enum member': 'enum'
};
var receivedTypesMap = {
    'a string': 'string',
    'a number': 'number',
    null: 'null',
    undefined: 'undefined',
    'a boolean': 'boolean',
    'an array': 'array',
    'an object': 'object',
    'a symbol': 'symbol',
    'a function': 'function',
    'a bigint': 'bigint'
};
var receivedTypesWithoutValue = [
    'null',
    'undefined',
    'boolean',
    'symbol',
    'function',
    'bigint'
];
/** Turns a string error message produced by a built-in purify codec into a meta object */
var parseError = function (error) {
    var oneOfCheck = error.match(oneofRegex);
    // One of the following problems occured: (0) *, (1) *
    if (oneOfCheck) {
        var remainer = error.replace(oneOfCheck[0], '');
        return {
            type: 'oneOf',
            errors: remainer
                .split(oneOfSeparatorRegex)
                .map(function (x) { return (0, exports.parseError)(x.replace(x.match(oneOfCounterRegex)[0], '')); })
        };
    }
    var failureCheck = error.match(failureRegex);
    // Expected an object, but received an array with value []
    if (failureCheck) {
        var receivedTypeRaw = error.split(failureCheck[2]).pop();
        var receivedType = receivedTypesMap[receivedTypeRaw.split(failureReceivedSeparator)[0]];
        if (receivedType) {
            var expectedTypeRaw = error
                .replace(failureCheck[1], '')
                .split(failureCheck[2])[0];
            return {
                type: 'failure',
                expectedType: expectedTypesMap[expectedTypeRaw],
                receivedType: receivedType,
                receivedValue: receivedTypesWithoutValue.includes(receivedType)
                    ? undefined
                    : JSON.parse(receivedTypeRaw.split(failureReceivedSeparator).pop())
            };
        }
    }
    // Problem with property "a": it does not exist in received object {}
    if (error.startsWith(missingPropertyMarker)) {
        var property = error.replace(missingPropertyMarker, '').split('": ')[0];
        return {
            type: 'property',
            property: property,
            error: {
                type: 'failure',
                receivedType: 'undefined'
            }
        };
    }
    // Problem with the value of property "a": *
    // Problem with key type of property "a": *
    if (error.startsWith(badPropertyMarker) ||
        error.startsWith(badPropertyKeyMarker)) {
        var _a = __read(error
            .replace(badPropertyMarker, '')
            .replace(badPropertyKeyMarker, '')
            .split(/": (.+)/)), property = _a[0], restOfError = _a.slice(1);
        return {
            type: 'property',
            property: property,
            error: (0, exports.parseError)(restOfError.join(''))
        };
    }
    // Problem with date string: *
    if (error.startsWith(dateFailureMarket)) {
        return (0, exports.parseError)(error.replace(dateFailureMarket, ''));
    }
    //  Problem with the value at index 0: *
    if (error.startsWith(indexMarker)) {
        var _b = __read(error
            .replace(indexMarker, '')
            .split(/: (.+)/)), index = _b[0], restOfError = _b.slice(1);
        return {
            type: 'index',
            index: Number(index),
            error: (0, exports.parseError)(restOfError.join(''))
        };
    }
    return { type: 'custom', message: error };
};
exports.parseError = parseError;
